File: c-quest04/ex05/my_spaceship.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
**
** QWASAR.IO -- my_spaceship
**
** @param {char*} flightPath
**
** @return {char*}
**
*/

#define FORMAT_STRING "{x: %i, y: %i, direction: '%s'}" 

#ifndef STRUCT_DIRECTION
#define STRUCT_DIRECTION
struct direction
{
    char* Up;
    char* Down;
    char* Left;
    char* Right; 
};
#endif

char* my_spaceship(char* flightPath)
{
    struct direction direct;
    direct.Up = "up";
    direct.Down = "down";
    direct.Left = "left";
    direct.Right = "right";

    int i;
    int X = 0, Y = 0; 
    char* buffer;
    char *currentDirection;
    currentDirection = direct.Up;
    buffer = malloc(200);

    for (i = 0; i < strlen(flightPath); i++) {
        if (flightPath[i] == 'R') {
            if(currentDirection == direct.Up){
                currentDirection = direct.Right;
            } else if(currentDirection == direct.Right) {
                currentDirection = direct.Down;
            } else if(currentDirection == direct.Down) {
                currentDirection = direct.Left;
            } else if(currentDirection == direct.Left) {
                currentDirection = direct.Up;
            }
        }
        else if (flightPath[i] == 'A') {
            if(currentDirection == direct.Up) {
                Y -= 1;
            } else if(currentDirection == direct.Right) {
                X += 1;
            } else if(currentDirection == direct.Down) {
                Y += 1;
            } else if(currentDirection == direct.Left) {
                X -= 1;
            }
        }
        else if (flightPath[i] == 'L') {
            if(currentDirection == direct.Up) {
                currentDirection = direct.Left;
            } else if(currentDirection == direct.Left) {
                currentDirection = direct.Down;
            } else if(currentDirection == direct.Down) {
                currentDirection = direct.Right;
            } else if(currentDirection == direct.Right) {
                currentDirection = direct.Up;
            }
        }
    }
    snprintf(buffer, 128, FORMAT_STRING, X, Y, currentDirection);
    return buffer;
}

File: my_spaceship/ex00/my_spaceship.js
function my_spaceship(flightPath) {
    var X = 0;
    var Y = 0;

    var Direction = {
          'Up': 'up',
          'Down': 'down',
          'Left': 'left',
          'Right': 'right'
    };

    var currentDirection = Direction['Up'];

    for (i = 0; i < flightPath.length; i++) {
        if (flightPath[i] == 'R') {
            if(currentDirection == Direction['Up']){
                currentDirection = Direction['Right'];
            } else if(currentDirection == Direction['Right']) {
                currentDirection = Direction['Down'];
            } else if(currentDirection == Direction['Down']) {
                currentDirection = Direction['Left'];
            } else if(currentDirection = Direction['Left']) {
                currentDirection = Direction['Up'];
            }
        }
        else if (flightPath[i] == 'A') {
            if(currentDirection == Direction['Up']) {
                Y -= 1;
            } else if(currentDirection == Direction['Right']) {
                X += 1;
            } else if(currentDirection == Direction['Down']) {
                Y += 1;
            } else if(currentDirection == Direction['Left']) {
                X -= 1;
            }
        }
        else if (flightPath[i] == 'L') {
            if(currentDirection == Direction['Up']) {
                currentDirection = Direction['Left'];
            } else if(currentDirection == Direction['Left']) {
                currentDirection = Direction['Down'];
            } else if(currentDirection == Direction['Down']) {
                currentDirection = Direction['Right'];
            } else if(currentDirection == Direction['Right']) {
                currentDirection = Direction['Up'];
            }
        }
    }
    movement = `{x: ${X}, y: ${Y}, direction: '${currentDirection}'}`;
    return movement;
}


// console.log(my_spaceship("RAALALL"));
// console.log(my_spaceship("AAAA"));
// console.log(my_spaceship(""));
// console.log(my_spaceship("RAARA"));
